api/collections/permissions.py

```diff
@@ -0,0 +1,78 @@
+from osf.models import AbstractNode, Collection, CollectedGuidMetadata
+
+class CreatorOrAdminOrPublic(permissions.BasePermission):
+    # Adapted from ContributorOrPublic
+    def has_object_permission(self, request, view, obj):
+        if isinstance(obj, CollectedGuidMetadata):
+            obj = obj.collection
+        if request.method in permissions.SAFE_METHODS:
+            if obj.is_public:
+                return True
+        auth = get_user_auth(request)
+        # TODO [IN-152]: Use django-guardian. For now, only creators
+        return obj.creator == auth.user
+
+class ReadOnlyIfCollectedRegistration(permissions.BasePermission):
+    """Makes PUT and POST forbidden for registrations."""
+    # Adapted from ReadOnlyIfRegistration
+    def has_object_permission(self, request, view, obj):
+        if isinstance(obj, AbstractNode) and obj.is_registration:
+            return request.method in permissions.SAFE_METHODS
+        return True
+
+class CreatorOrAdminOrPublicForPointers(permissions.BasePermission):
+    # Adapted from ContributorOrPublicForPointers
+    # Will only work for refs that point to AbstractNodes/Collections
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (CollectedGuidMetadata, Collection)), 'obj must be an Collection or CollectedGuidMetadata, got {}'.format(obj)
+        auth = get_user_auth(request)
+        collection = Collection.load(request.parser_context['kwargs']['node_id'])
+        pointer_node = collection.collectedguidmetadata_set.get(guid___id=request.parser_context['kwargs']['node_link_id']).guid.referent
+        if request.method in permissions.SAFE_METHODS:
+            has_collection_auth = collection.creator == auth.user
+            if isinstance(pointer_node, AbstractNode):
+                has_pointer_auth = pointer_node.can_view(auth)
+            elif isinstance(pointer_node, Collection):
+                has_pointer_auth = pointer_node.creator == auth.user
+            public = pointer_node.is_public
+            has_auth = public or (has_collection_auth and has_pointer_auth)
+            return has_auth
+        else:
+            return collection.creator == auth.user
+
+class CreatorOrAdminOrPublicForRelationshipPointers(permissions.BasePermission):
+    # Adapted from ContributorOrPublicForRelationshipPointers
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, dict)
+        auth = get_user_auth(request)
+        collection = obj['self']
+        has_collection_auth = collection.creator == auth.user
+
+        if request.method in permissions.SAFE_METHODS:
+            if collection.is_public:
+                return True
+        elif request.method == 'DELETE':
+            return has_collection_auth
+
+        if not has_collection_auth:
+            return False
+        pointer_nodes = []
+        for pointer in request.data.get('data', []):
+            node = AbstractNode.load(pointer['id'])
```

Comment:
Is there anything that prevents linking quickfilesnodes? I saw a few places reference it, but just wanted to check. Or is that allowed?

Reply:
It probably shouldn't be allowed, but currently is. Verified this on staging by `curl`-ing my QFN into a collection:
```
(osf) ~/D/osf.io ❯ curl 'https://staging-api.osf.io/v2/collections/<collection.id>/relationships/linked_nodes/?<params>' <some args> --data-binary '{"data":[{"type":"linked_nodes","id":"<QFN.id>"}]}'
```
Then checked postgres to verify:
```
>>> c = Collection.load('<id>')
>>> c.linked_nodes
<AbstractNodeQuerySet [<other nodes>, <QuickFilesNode: (title=u"\u041c\u044d\u0442\u044c\u044e \u0424\u0440\u0435\u0439\u0437\xeb\u0440's Quick Files", category=u'') with <id>'> <more other nodes>]>
```